indra/newview/llfloateremojipicker.cpp

```diff
@@ -631,9 +674,186 @@ BOOL LLFloaterEmojiPicker::handleKeyHere(KEY key, MASK mask)
+            }
+            break;
+        }
+        itf++;
+    }
+    // Append new if not found
+    if (itf == sFrequentlyUsed.end())
+        sFrequentlyUsed.push_back(std::make_pair(emoji, 1));
+}
+
+void LLFloaterEmojiPicker::loadState()
+{
+    if (!sStateFileName.empty())
+        return; // Already loaded
+
+    sStateFileName = gDirUtilp->getExpandedFilename(LL_PATH_USER_SETTINGS, "emoji_floater_state.xml");
+
+    llifstream file;
+    file.open(sStateFileName.c_str());
+    if (!file.is_open())
+    {
+        LL_WARNS() << "Emoji floater state file is missing or inaccessible: " << sStateFileName << LL_ENDL;
+        return;
+    }
+
+    LLSD state;
+    LLSDSerialize::fromXML(state, file);
+    if (state.isUndefined())
+    {
+        LL_WARNS() << "Emoji floater state file is missing or ill-formed: " << sStateFileName << LL_ENDL;
+        return;
+    }
+
+    sSelectedGroupIndex = state[sKeySelectedGroupIndex].asInteger();
+
+    sSearchPattern = state[sKeySearchPattern].asString();
+
+    // Load and parse sRecentlyUsed
+    std::string recentlyUsed = state[sKeyRecentlyUsed];
+    std::vector<std::string> rtokens = LLStringUtil::getTokens(recentlyUsed, ",");
+    int maxCountR = 20;
+    for (const std::string& token : rtokens)
+    {
+        llwchar emoji = (llwchar)atoi(token.c_str());
+        if (std::find(sRecentlyUsed.begin(), sRecentlyUsed.end(), emoji) == sRecentlyUsed.end())
+        {
+            sRecentlyUsed.push_back(emoji);
+            if (!--maxCountR)
+                break;
+        }
+    }
+
+    // Load and parse sFrequentlyUsed
+    std::string frequentlyUsed = state[sKeyFrequentlyUsed];
+    std::vector<std::string> ftokens = LLStringUtil::getTokens(frequentlyUsed, ",");
+    int maxCountF = 20;
+    for (const std::string& token : ftokens)
+    {
+        std::vector<std::string> pair = LLStringUtil::getTokens(token, ":");
+        if (pair.size() == 2)
+        {
+            llwchar emoji = (llwchar)atoi(pair[0].c_str());
+            if (emoji)
+            {
+                U32 count = atoi(pair[1].c_str());
+                auto it = std::find_if(sFrequentlyUsed.begin(), sFrequentlyUsed.end(),
+                    [emoji](std::pair<llwchar, U32>& it) { return it.first == emoji; });
+                if (it != sFrequentlyUsed.end())
+                {
+                    it->second += count;
+                }
+                else
+                {
+                    sFrequentlyUsed.push_back(std::make_pair(emoji, count));
+                    if (!--maxCountF)
+                        break;
+                }
+            }
+        }
+    }
+
+    // Normalize by minimum
+    if (!sFrequentlyUsed.empty())
+    {
+        U32 delta = sFrequentlyUsed.back().second;
+        for (auto& it : sFrequentlyUsed)
+        {
+            it.second = std::max((U32)0, it.second - delta);
+        }
+    }
+}
+
+void LLFloaterEmojiPicker::saveState()
+{
+    if (sStateFileName.empty())
```

Comment:
I suggest checking mSecondInstance. Generally we treat files from under second instance as 'read only' to prevent issues.

Reply:
mSecondInstance relates to another opera

the floater is created each time it appears on the screen
but static data members preserve their values even after the floater is destroyed
that's why it is enough to load the saved state only once, on the 1st creating of the floater
the state is saved each time the floater disappears, but the state stays loaded in the memory
so no needs to load it again

Reply:
>on the 1st creating of the floater the state is saved each time the floater disappears
Second instance does not save (or is not supposed to) library cache, asset or texture cache due to potential overlap with main instance. Probably shouldn't even save chrome cache. And I probably should do the same for keybindings...

Example: if both instances of viewer were logged into the same region that restarted/logged out both agents, both viewers will try to save to file simultaneously, which might cause a mess since you didn't add write protections.

I think you should either check for second instance here; instead of LL_PATH_USER_SETTINGS use LL_PATH_PER_SL_ACCOUNT; or lock files before writing.
I think first option is preferable, third one not practical (due to needing OS specific lock methods)